tests/combine_span_test.cpp

```diff
@@ -0,0 +1,280 @@
+  logp
+ */
+nuts::SpanW<double> dummy_span(double logp) {
+  Eigen::VectorXd theta = Eigen::VectorXd::Random(2);
+  Eigen::VectorXd rho = Eigen::VectorXd::Random(2);
+  Eigen::VectorXd grad = Eigen::VectorXd::Random(2);
+
+  nuts::SpanW<double> span(std::move(theta), std::move(rho), std::move(grad),
+                           logp);
+
+  span.logp_bk_ = -logp;
+
+  span.theta_bk_ = Eigen::VectorXd::Random(2);
+  span.rho_bk_ = Eigen::VectorXd::Random(2);
+  span.grad_theta_bk_ = Eigen::VectorXd::Random(2);
+
+  span.logp_fw_ = logp + 10;
+
+  span.theta_fw_ = Eigen::VectorXd::Random(2);
+  span.rho_fw_ = Eigen::VectorXd::Random(2);
+  span.grad_theta_fw_ = Eigen::VectorXd::Random(2);
+
+  return span;
+}
+
+/**  Mocked RNG that returns a set value deterministically */
+struct mock_rng {
+  double value;
+  double uniform_real_01() { return value; }
+};
+
+// helpers to use GoogleTest's parameterization system
+using Forwards =
+    std::integral_constant<nuts::Direction, nuts::Direction::Forward>;
+using Backwards =
+    std::integral_constant<nuts::Direction, nuts::Direction::Backward>;
+using Metropolis =
+    std::integral_constant<nuts::Update, nuts::Update::Metropolis>;
+using Barker = std::integral_constant<nuts::Update, nuts::Update::Barker>;
+
+using ForwardsMetropolis = std::tuple<Forwards, Metropolis>;
+using BackwardsMetropolis = std::tuple<Backwards, Metropolis>;
+using ForwardsBarker = std::tuple<Forwards, Barker>;
+using BackwardsBarker = std::tuple<Backwards, Barker>;
+
+}  // namespace span_test
+
+// macros to compare parts of Spans
+#define EXPECT_FORWARD_ENDPOINT_EQUAL(span_recieved, span_expected)      \
+  EXPECT_EQ(span_recieved.theta_fw_, span_expected.theta_fw_);           \
+  EXPECT_EQ(span_recieved.rho_fw_, span_expected.rho_fw_);               \
+  EXPECT_EQ(span_recieved.grad_theta_fw_, span_expected.grad_theta_fw_); \
+  EXPECT_EQ(span_recieved.logp_fw_, span_expected.logp_fw_);
+
+#define EXPECT_BACKWARD_ENDPOINT_EQUAL(span_recieved, span_expected)     \
+  EXPECT_EQ(span_recieved.theta_bk_, span_expected.theta_bk_);           \
+  EXPECT_EQ(span_recieved.rho_bk_, span_expected.rho_bk_);               \
+  EXPECT_EQ(span_recieved.grad_theta_bk_, span_expected.grad_theta_bk_); \
+  EXPECT_EQ(span_recieved.logp_bk_, span_expected.logp_bk_);
+
+#define EXPECT_SELECTED_POINT_EQUAL(span_recieved, input, new_logp) \
+  EXPECT_EQ(span_recieved.theta_select_, input.theta_select_);      \
+  EXPECT_EQ(span_recieved.grad_select_, input.grad_select_);        \
+  EXPECT_FLOAT_EQ(span_recieved.logp_, new_logp);
+
+// The forwards/backwards pieces of the new span are update agnostic
+template <typename T>
+class CombineSpansUpdateAgnostic : public ::testing::Test {};
+
+using BothUpdates = ::testing::Types<span_test::Metropolis, span_test::Barker>;
+TYPED_TEST_SUITE(CombineSpansUpdateAgnostic, BothUpdates);
+
+TYPED_TEST(CombineSpansUpdateAgnostic, ForwardEndpointsCorrect) {
+  constexpr auto update = TypeParam::value;
```

Comment:
Where is `TypeParam` coming from?  Is it a Google test constant?  I don't see it defined anywhere.

Reply:
How is `TYPED_TEST` different from a regular test?  I can't understand what's going on here.